### ISSUE TITLE
[Bug] interaction of listening to backspace delete + calling clear() in unfinished drawing state results in errors

### ISSUE THREAD

**Issue by bdon** [JamesLMilner/terra-draw#304]
**Describe the bug**
When a `RectangleMode` rectangle drawing has started, but not finished, and `clear()` is called before finishing, the store ends up in an error state. 

**Terra Draw npm version**
1.0.0-beta.0

**To Reproduce**

The adaption of the basic example below is designed to allow for a single rectangle selection at a time. Deleting the current rectangle can be done with SelectMode + Backspace, or by using a button. 

The error appears when the rectangle mode is started but not yet finished, and the button is clicked in the intermediate state. 

```html
<!doctype html>
<html>
  <head>
    <link
      rel="stylesheet"
      href="https://unpkg.com/maplibre-gl@3.6.2/dist/maplibre-gl.css"
    />
    <script src="https://unpkg.com/maplibre-gl@3.6.2/dist/maplibre-gl.js"></script>
    <script src="https://unpkg.com/terra-draw@1.0.0-beta.0/dist/terra-draw.umd.js"></script>
  </head>
  <body>
    <div id="map" style="height:420px;"></div>
    Demo for an app that allows only one rectangle selection at a time. the button or select+backspace clears the drawing.

    click the map once to start drawing a rectangle, but before finishing
    <button id="button">Click Me</button>
    <script>
      var map = new maplibregl.Map({
        container: "map",
        style: "https://demotiles.maplibre.org/style.json",
        center: [-4.49049, 54.189649], // [lng, lat]
        zoom: 5,
      });

      const draw = new terraDraw.TerraDraw({
        adapter: new terraDraw.TerraDrawMapLibreGLAdapter({
          map: map,
          lib: maplibregl,
        }),
        modes: [new terraDraw.TerraDrawRectangleMode(), new terraDraw.TerraDrawSelectMode({
          keyEvents: {
            delete: "Backspace"
          },
          styles: {
              selectedPolygonOutlineColor: "red",
          },
          flags: {
            rectangle: {
              deletable: true,
              feature: {
                coordinates: {
                },
              }
            }
          }
        })],
      });

      draw.start();
      draw.setMode("rectangle");

      draw.on("finish", () => {
        console.log("onFinish");
        draw.setMode("select");
      })

      draw.on("change", (x, type) => {
        console.log("onChange", x, type);
        if (type === "delete") {
          // if there are now 0 features, because of deletable (backspace on keystroke), go back into rectangle mode.
          const features = draw.getSnapshot();
          if (features.length === 0) {
            draw.setMode("rectangle");
          }
        }
      });

      document.getElementById("button").onclick = () => {
        console.log("Clicked!");
        draw.clear();
        draw.setMode("rectangle");
      }
    </script>

  </body>
</html>
```

**Expected behavior**
`clear` and `setMode` should be able to be called from any state to reset. 

**Screenshots**
Video:

https://github.com/user-attachments/assets/7af417ef-b1c2-4a8e-a7ab-e67232cc40f5

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Safari
 - Version 17.5

**Additional context**

**Comment by bdon** [comment]
I think the issue is more directly related to the way I'm listening for the backspace delete event here:

```js
if (type === "delete") {
          // if there are now 0 features, because of deletable (backspace on keystroke), go back into rectangle mode.
          const features = draw.getSnapshot();
          if (features.length === 0) {
            draw.setMode("rectangle");
          }
        }
```

This seems to interact with `clear` in an unexpected way. If I comment out that event handler the `clear()` in an intermediate drawing state works fine.

**Comment by JamesLMilner** [comment]
@bdon I can reproduce this and think I figured out why it is happening.

Because Terra Draw is synchronous, the change event is triggered immediately by the store and the code path is followed, however this is before `this.currentRectangleId = undefined;` has run and when we run `draw.setMode` in the `change` callback, the original `clear` call from the button click has not finished and `this.currentRectangleId` is still defined.

The fix for this is to clean up the state, _before_ we trigger any store updates which may have listeners. I'll raise a PR!

**Comment by bdon** [comment]
Confirmed fixed in `1.0.0-beta.1`, thanks!
